Sources/Flush.swift

```diff
@@ -86,53 +89,47 @@ class Flush: AppLifecycle {
-            if let requestData = requestData {
-                let semaphore = DispatchSemaphore(value: 0)
-                #if os(iOS)
+        flushBatchQueue.sync {
+            var mutableQueue = queue
+            while !mutableQueue.isEmpty {
+                let batchSize = min(mutableQueue.count, APIConstants.batchSize)
+                let range = 0..<batchSize
+                let batch = Array(mutableQueue[range])
+                let ids: [Int32] = batch.map { entity in
+                    (entity["id"] as? Int32) ?? 0
+                }
+                // Log data payload sent
+                Logger.debug(message: "Sending batch of data")
+                Logger.debug(message: batch as Any)
+                let requestData = JSONHandler.encodeAPIData(batch)
+                if let requestData = requestData {
+                    #if os(iOS)
                     if !MixpanelInstance.isiOSAppExtension() {
                         delegate?.updateNetworkActivityIndicator(true)
                     }
-                #endif // os(iOS)
-                flushRequest.sendRequest(requestData,
-                                         type: type,
-                                         useIP: useIPAddressForGeoLocation,
-                                         completion: { [weak self, semaphore] success in
-                                            guard let self = self else { return }
-                                            #if os(iOS)
-                                                if !MixpanelInstance.isiOSAppExtension() {
-                                                    self.delegate?.updateNetworkActivityIndicator(false)
-                                                }
-                                            #endif // os(iOS)
-                                            if success {
-                                                // remove
-                                                self.delegate?.flushSuccess(type: type, ids: ids)
-                                                mutableQueue = self.removeProcessedBatch(batchSize: batchSize,
-                                                                                         queue: mutableQueue,
-                                                                                         type: type)
-                                            }
-                                            shouldContinue = success
-                                            semaphore.signal()
-                })
-                _ = semaphore.wait(timeout: DispatchTime.distantFuture)
-            }
-            if !shouldContinue {
-                break
+                    #endif // os(iOS)
+                    let success = flushRequest.sendRequest(requestData,
+                                                           type: type,
+                                                           useIP: useIPAddressForGeoLocation)
```

Comment:
move the semaphore into `sendRequest` and make it synchronous from the perspective of `flushQueueInBatches`